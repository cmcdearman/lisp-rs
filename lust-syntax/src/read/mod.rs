pub mod sexpr;
pub mod token;

use self::{
    sexpr::{Atom, AtomKind, DataList, Lit, Root, Sexpr, SexprKind, SynList},
    token::Token,
};
use chumsky::{
    extra,
    input::{Stream, ValueInput},
    prelude::{Input, Rich},
    primitive::just,
    recursive::recursive,
    select, IterParser, Parser,
};
use logos::Logos;
use lust_utils::{intern::InternedString, list::List, span::Span};

#[derive(Debug, Clone, PartialEq)]
pub enum SyntaxError<'a> {
    LexError(Span),
    ParseError(Rich<'a, Token, Span, &'a str>),
}

pub fn read<'src>(src: &'src str) -> (Option<Root>, Vec<SyntaxError<'src>>) {
    let mut errs = Vec::new();
    let mut tokens = vec![];
    for (res, span) in Token::lexer(src).spanned() {
        match res {
            Ok(tok) => tokens.push((tok, Span::from(span))),
            Err(_) => {
                errs.push(SyntaxError::LexError(Span::from(span.clone())));
                tokens.push((Token::Error, Span::from(span)))
            }
        }
    }
    if !errs.is_empty() {
        return (None, errs);
    }
    let tok_stream = Stream::from_iter(tokens).spanned(Span::from(src.len()..src.len()));
    let (root, errs) = root_reader().parse(tok_stream).into_output_errors();
    (
        root,
        errs.into_iter()
            .map(|err| SyntaxError::ParseError(err))
            .collect(),
    )
}

fn root_reader<'a, I: ValueInput<'a, Token = Token, Span = Span>>(
) -> impl Parser<'a, I, Root, extra::Err<Rich<'a, Token, Span>>> {
    sexpr_reader()
        .repeated()
        .collect()
        .map_with_span(Root::new)
        .boxed()
}

fn sexpr_reader<'a, I: ValueInput<'a, Token = Token, Span = Span>>(
) -> impl Parser<'a, I, Sexpr, extra::Err<Rich<'a, Token, Span>>> {
    recursive(|sexpr| {
        let atom = ident_reader()
            .map(AtomKind::Sym)
            .or(lit_reader().map(AtomKind::Lit))
            .map_with_span(Atom::new)
            .map(SexprKind::Atom)
            .map_with_span(Sexpr::new)
            .boxed();

        let syn_list = sexpr
            .clone()
            .repeated()
            .at_least(1)
            .collect::<Vec<_>>()
            .map(List::from)
            .map_with_span(SynList::new)
            .map(SexprKind::SynList)
            .map_with_span(Sexpr::new)
            .delimited_by(just(Token::LParen), just(Token::RParen));

        let data_list = sexpr
            .clone()
            .repeated()
            .at_least(1)
            .collect::<Vec<_>>()
            .map(List::from)
            .map_with_span(DataList::new)
            .map(SexprKind::DataList)
            .map_with_span(Sexpr::new)
            .delimited_by(just(Token::LBrack), just(Token::RBrack));

        let empty = just(Token::LBrack)
            .then(just(Token::RBrack))
            .map_with_span(|_, span| SexprKind::DataList(DataList::new(List::Empty, span)))
            .map_with_span(Sexpr::new);

        let vector = sexpr
            .clone()
            .repeated()
            .collect()
            .map(SexprKind::Vector)
            .map_with_span(Sexpr::new)
            .delimited_by(just(Token::HashLBrack), just(Token::RBrack));

        // quote = "'" sexpr
        let quote = just(Token::Quote)
            .map_with_span(|_, span| span)
            .then(sexpr.clone())
            .map(|(span, sexpr)| {
                let mut list = List::Empty;
                list.push_front(sexpr);
                list.push_front(Sexpr::new(
                    SexprKind::Atom(Atom::new(
                        AtomKind::Sym(InternedString::from("quote")),
                        span,
                    )),
                    span,
                ));
                SexprKind::SynList(SynList::new(list, span))
            })
            .map_with_span(Sexpr::new);

        let quasiquote = just(Token::Backquote)
            .map_with_span(|_, span| span)
            .then(sexpr.clone())
            .map(|(span, sexpr)| {
                let mut list = List::Empty;
                list.push_front(sexpr);
                list.push_front(Sexpr::new(
                    SexprKind::Atom(Atom::new(
                        AtomKind::Sym(InternedString::from("quasiquote")),
                        span,
                    )),
                    span,
                ));
                SexprKind::SynList(SynList::new(list, span))
            })
            .map_with_span(Sexpr::new);

        let unquote = just(Token::Comma)
            .map_with_span(|_, span| span)
            .then(sexpr.clone())
            .map(|(span, sexpr)| {
                let mut list = List::Empty;
                list.push_front(sexpr);
                list.push_front(Sexpr::new(
                    SexprKind::Atom(Atom::new(
                        AtomKind::Sym(InternedString::from("unquote")),
                        span,
                    )),
                    span,
                ));
                SexprKind::SynList(SynList::new(list, span))
            })
            .map_with_span(Sexpr::new);

        let unquote_splice = just(Token::CommaAt)
            .map_with_span(|_, span| span)
            .then(sexpr.clone())
            .map(|(span, sexpr)| {
                let mut list = List::Empty;
                list.push_front(sexpr);
                list.push_front(Sexpr::new(
                    SexprKind::Atom(Atom::new(
                        AtomKind::Sym(InternedString::from("unquote-splicing")),
                        span,
                    )),
                    span,
                ));
                SexprKind::SynList(SynList::new(list, span))
            })
            .map_with_span(Sexpr::new);

        atom.or(syn_list)
            .or(data_list)
            .or(empty)
            .or(vector)
            .or(quote)
            .or(quasiquote)
            .or(unquote)
            .or(unquote_splice)
    })
}

fn ident_reader<'a, I: ValueInput<'a, Token = Token, Span = Span>>(
) -> impl Parser<'a, I, InternedString, extra::Err<Rich<'a, Token, Span>>> {
    select! {
        Token::Ident(name) => name,
    }
}

fn lit_reader<'a, I: ValueInput<'a, Token = Token, Span = Span>>(
) -> impl Parser<'a, I, Lit, extra::Err<Rich<'a, Token, Span>>> {
    select! {
        Token::Int(n) => Lit::Int(n),
        Token::Float(n) => Lit::Float(n),
        Token::Rational(n) => Lit::Rational(n),
        Token::String(s) => Lit::Str(s),
    }
}
