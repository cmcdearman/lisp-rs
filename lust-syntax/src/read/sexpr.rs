use lust_utils::{
    intern::InternedString,
    list::List,
    num::{BigInt, BigRational, Float},
    span::Span,
};
use std::fmt::Display;

#[derive(Debug, Clone, PartialEq)]
pub struct Root {
    sexprs: Vec<Sexpr>,
    span: Span,
}

impl Root {
    pub fn new(sexprs: Vec<Sexpr>, span: Span) -> Self {
        Self { sexprs, span }
    }

    pub fn sexprs(&self) -> &[Sexpr] {
        &self.sexprs
    }
}

impl Display for Root {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        for s in &self.sexprs {
            writeln!(f, "{}", s)?;
        }
        Ok(())
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Sexpr {
    kind: Box<SexprKind>,
    span: Span,
}

impl Sexpr {
    pub fn new(kind: SexprKind, span: Span) -> Self {
        Self {
            kind: Box::new(kind),
            span,
        }
    }

    pub fn kind(&self) -> &SexprKind {
        &self.kind
    }

    pub fn span(&self) -> &Span {
        &self.span
    }

    pub fn replace(&mut self, kind: SexprKind) {
        self.kind = Box::new(kind);
    }

    pub fn replace_sym(&mut self, sym: InternedString) {
        // recursively replace all instances of the symbol
        match self.kind() {
            SexprKind::Atom(a) => match a.kind() {
                AtomKind::Sym(s) => {
                    if s == &sym {
                        *self = Sexpr::new(
                            SexprKind::Atom(Atom::new(AtomKind::Sym(sym), *a.span())),
                            *self.span(),
                        );
                    }
                }
                _ => (),
            },
            SexprKind::SynList(l) => {
                for s in l.list().iter() {
                    s.replace_sym(sym.clone());
                }
            }
            SexprKind::DataList(l) => {
                for s in l.list().iter() {
                    s.replace_sym(sym.clone());
                }
            }
            SexprKind::Vector(v) => {
                for s in v {
                    s.replace_sym(sym.clone());
                }
            }
        }
    }
}

impl Display for Sexpr {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.kind)
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum SexprKind {
    Atom(Atom),
    SynList(SynList),
    DataList(DataList),
    Vector(Vec<Sexpr>),
}

impl Display for SexprKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SexprKind::Atom(a) => write!(f, "{}", a),
            SexprKind::SynList(l) => write!(f, "{}", l),
            SexprKind::DataList(l) => write!(f, "{}", l),
            SexprKind::Vector(v) => {
                write!(f, "[")?;
                for (i, s) in v.iter().enumerate() {
                    if i != 0 {
                        write!(f, " ")?;
                    }
                    write!(f, "{}", s)?;
                }
                write!(f, "]")
            }
        }
    }
}

#[derive(Debug, Clone, PartialEq, Default)]
pub struct SynList {
    list: List<Sexpr>,
    span: Span,
}

impl SynList {
    pub fn new(list: List<Sexpr>, span: Span) -> Self {
        Self { list, span }
    }

    pub fn list(&self) -> &List<Sexpr> {
        &self.list
    }

    pub fn span(&self) -> &Span {
        &self.span
    }

    pub fn head(&self) -> Option<&Sexpr> {
        self.list.head()
    }

    pub fn tail(&self) -> Option<&List<Sexpr>> {
        self.list.tail()
    }
}

impl Display for SynList {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.list)
    }
}

#[derive(Debug, Clone, PartialEq, Default)]
pub struct DataList {
    list: List<Sexpr>,
    span: Span,
}

impl DataList {
    pub fn new(list: List<Sexpr>, span: Span) -> Self {
        Self { list, span }
    }

    pub fn list(&self) -> &List<Sexpr> {
        &self.list
    }

    pub fn span(&self) -> &Span {
        &self.span
    }

    pub fn head(&self) -> Option<&Sexpr> {
        self.list.head()
    }

    pub fn tail(&self) -> Option<&List<Sexpr>> {
        self.list.tail()
    }
}

impl Display for DataList {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "[")?;
        for (i, s) in self.list.iter().enumerate() {
            if i != 0 {
                write!(f, " ")?;
            }
            write!(f, "{}", s)?;
        }
        write!(f, "]")
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Atom {
    kind: Box<AtomKind>,
    span: Span,
}

impl Atom {
    pub fn new(kind: AtomKind, span: Span) -> Self {
        Self {
            kind: Box::new(kind),
            span,
        }
    }

    pub fn kind(&self) -> &AtomKind {
        &self.kind
    }

    pub fn span(&self) -> &Span {
        &self.span
    }
}

impl Display for Atom {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.kind)
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum AtomKind {
    Sym(InternedString),
    Lit(Lit),
}

impl Display for AtomKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            AtomKind::Sym(s) => write!(f, "{}", s),
            AtomKind::Lit(l) => write!(f, "{}", l),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Lit {
    Int(BigInt),
    Float(Float),
    Rational(BigRational),
    Str(InternedString),
    Bool(bool),
    Char(char),
}

impl Display for Lit {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Lit::Int(i) => write!(f, "{}", i),
            Lit::Float(fl) => write!(f, "{}", fl),
            Lit::Rational(r) => write!(f, "{}", r),
            Lit::Str(s) => write!(f, "{}", s),
            Lit::Bool(b) => write!(f, "{}", b),
            Lit::Char(c) => write!(f, "{}", c),
        }
    }
}
