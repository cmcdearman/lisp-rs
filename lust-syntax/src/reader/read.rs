use super::{
    error::{ReadResult, ReaderError},
    sexpr::{Root, Sexpr},
    token::{Token, TokenKind},
};
use logos::{Lexer, Logos};
use lust_util::{
    list::List,
    span::{Span, Spannable, Spanned},
};

pub struct Reader<'src> {
    src: &'src str,
    lexer: Lexer<'src, TokenKind>,
    peek: Option<Token>,
    errors: Vec<ReaderError>,
}

impl<'src> Reader<'src> {
    /// Creates a new [`Parser`].
    pub fn new(src: &'src str) -> Self {
        Self {
            src,
            lexer: TokenKind::lexer(src),
            peek: None,
            errors: vec![],
        }
    }

    fn fetch_token(&mut self) -> Token {
        match self.lexer.next().map(|res| (res, self.lexer.span())) {
            Some((res, s)) => match res {
                Ok(t) => t.spanned(s.into()),
                Err(_) => {
                    self.errors.push(SyntaxError::LexerError.spanned(s.into()));
                    self.fetch_token()
                }
            },
            None => TokenKind::Eof.spanned(self.lexer.span().into()),
        }
    }

    fn peek(&mut self) -> Token {
        if let Some(token) = self.peek.clone() {
            token
        } else {
            let token = self.fetch_token();
            self.peek = Some(token.clone());
            token
        }
    }

    fn next(&mut self) -> Token {
        if let Some(token) = self.peek.take() {
            token
        } else {
            self.fetch_token()
        }
    }

    fn at(&mut self, kind: TokenKind) -> bool {
        self.peek().value == kind
    }

    fn eat(&mut self, kind: TokenKind) -> ReadResult<()> {
        if self.at(kind) {
            self.next();
            Ok(())
        } else {
            let next = self.next();
            Err(ReaderError::UnexpectedToken(next))
        }
    }

    fn text(&self) -> &'src str {
        &self.src[self.lexer.span()]
    }

    pub fn parse(&mut self) -> (Root, Vec<ReaderError>) {
        let mut sexprs = vec![];
        while !self.at(TokenKind::Eof) {
            if let Ok(s) = self.sexpr() {
                sexprs.push(s);
            } else {
                self.errors.push(ReaderError::UnexpectedToken(self.peek()));
                self.next();
            }
        }
        (Root { sexprs }, self.errors.clone())
    }

    fn sexpr(&mut self) -> ReadResult<Spanned<Sexpr>> {
        match self.peek().0 {
            TokenKind::LParen => self.list(),
            _ => self.atom(),
        }
    }

    fn list(&mut self) -> ReadResult<Spanned<Sexpr>> {
        let start = self.peek().span;
        if !self.eat(TokenKind::LParen)? {
            return Err(ReaderError::UnmatchedParen(self.peek().1));
        }
        let mut sexprs = vec![];
        while !self.at(&Token::RParen) {
            let s = self.sexpr()?;
            sexprs.push(s);
            self.next();
        }
        if !self.eat(&Token::RParen) {
            return Err(ReaderError::UnmatchedParen(self.peek().1));
        }
        let list: List<Spanned<Sexpr>> = sexprs.into_iter().rev().collect();
        let end = self.peek().1.end as usize;
        Ok((Sexpr::Cons(Box::new(list)), Span::from(start..end)))
    }

    fn atom(&mut self) -> ReadResult<Spanned<Sexpr>> {
        match self.peek().0 {
            Token::Int(i) => Ok((Sexpr::Atom(Atom::Lit(Lit::Int(i))), self.peek().1)),
            Token::Rational(r) => Ok((Sexpr::Atom(Atom::Lit(Lit::Rational(r))), self.peek().1)),
            Token::Real(r) => Ok((Sexpr::Atom(Atom::Lit(Lit::Real(r))), self.peek().1)),
            Token::Char(c) => Ok((Sexpr::Atom(Atom::Lit(Lit::Char(c))), self.peek().1)),
            Token::String(s) => Ok((
                Sexpr::Atom(Atom::Lit(Lit::String(InternedString::from(
                    &s[1..(s.len() - 1)],
                )))),
                self.peek().1,
            )),
            Token::Ident(name) => Ok((
                Sexpr::Atom(Atom::Symbol(InternedString::from(name))),
                self.peek().1,
            )),
            _ => Err(ReaderError::UnexpectedToken(self.peek())),
        }
    }
}
