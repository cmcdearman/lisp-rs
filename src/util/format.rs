#[derive(Clone, PartialEq)]
pub struct Format<T> {
    pub indent: usize,
    pub value: T,
}

impl<T> Format<T> {
    pub fn new(indent: usize, value: T) -> Self {
        Self { indent, value }
    }
}

pub fn spaces(n: usize) -> String {
    " ".repeat(n)
}
