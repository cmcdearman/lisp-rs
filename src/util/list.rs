use chumsky::container::Container;
use itertools::join;
use std::fmt::Display;

/// A singly-linked list with owned nodes.
#[derive(Debug, Clone, Default, PartialEq)]
pub enum List<T> {
    Node(Box<(T, List<T>)>),
    // Node {
    //     car: Box<T>,
    //     cdr: Box<List<T>>,
    // },
    #[default]
    Nil,
}

impl<T> List<T> {
    pub fn new(car: T, cdr: List<T>) -> Self {
        Self::Node(Box::new((car, cdr)))
    }
}

// #[derive(Debug, Clone, PartialEq)]
// pub struct ConsPair<T> {
//     pub car: T,
//     pub cdr: ConsPair<T>,
// }

impl<T> Display for List<T>
where
    T: Display + Clone,
{
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            List::Node(_) => {
                write!(f, "[")?;
                write!(f, "{}", join(self.clone(), ", "))?;
                write!(f, "]")
            }
            List::Nil => write!(f, "[]"),
        }
    }
}

impl<T: Clone> Iterator for List<T> {
    type Item = T;

    fn next(&mut self) -> Option<Self::Item> {
        match self {
            List::Node(n) => {
                let (data, next) = n.as_ref();
                let data = data.clone();
                *self = next.clone();
                Some(data)
            }
            List::Nil => None,
        }
    }
}

impl<T: Clone> ExactSizeIterator for List<T> {
    fn len(&self) -> usize {
        self.clone().count()
    }
}

impl<T> FromIterator<T> for List<T> {
    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
        iter.into_iter()
            .fold(Self::Nil, |list, data| Self::Node(Box::new((data, list))))
    }
}

impl<T: Clone> DoubleEndedIterator for List<T> {
    fn next_back(&mut self) -> Option<Self::Item> {
        match self {
            List::Node(n) => {
                let (data, next) = n.as_ref();
                let data = data.clone();
                *self = next.clone();
                Some(data)
            }
            List::Nil => None,
        }
    }
}

impl<T: Clone + Default> Container<T> for List<T> {
    fn push(&mut self, item: T) {
        *self = Self::Node(Box::new((item, self.clone())));
    }

    fn with_capacity(n: usize) -> Self {
        let mut list = Self::Nil;
        for _ in 0..n {
            list = Self::Node(Box::new((Default::default(), list)));
        }
        list
    }
}
