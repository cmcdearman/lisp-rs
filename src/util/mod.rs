pub mod format;
pub mod intern;
pub mod list;
pub mod meta;
pub mod node;
pub mod span;
pub mod unique_id;
