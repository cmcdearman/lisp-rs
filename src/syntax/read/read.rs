use super::{
    sexpr::{Atom, Root},
    token::Token,
};
use crate::util::{intern::InternedString, node::SrcNode, span::Span};
use chumsky::{
    extra,
    input::{Stream, ValueInput},
    prelude::{Input, Rich},
    select, IterParser, Parser,
};
use logos::Logos;

pub type ReadError<'a> = extra::Err<Rich<'a, Token, Span>>;

fn ident_reader<'a, I: ValueInput<'a, Token = Token, Span = Span>>(
) -> impl Parser<'a, I, InternedString, extra::Err<Rich<'a, Token, Span>>> {
    select! {
        Token::Symbol(name) => InternedString::from(name),
    }
}

fn atom_reader<'a, I: ValueInput<'a, Token = Token, Span = Span>>(
) -> impl Parser<'a, I, SrcNode<Atom>, extra::Err<Rich<'a, Token, Span>>> {
    select! {
        Token::Symbol(name) => Atom::Symbol(InternedString::from(name)),
        Token::Number(n) => Atom::Number(n),
        Token::String(s) => Atom::String(s),
    }
}

fn reader<'a, I: ValueInput<'a, Token = Token, Span = Span>>(
) -> impl Parser<'a, I, SrcNode<Root>, extra::Err<Rich<'a, Token, Span>>> {
    atom_reader()
        .repeated()
        .collect()
        .map(|sexprs| Root { sexprs })
}

pub fn read<'src>(src: &'src str) -> (Option<Root>, Vec<ReadError<'src>>) {
    let tokens = Token::lexer(&src).spanned().map(|(tok, span)| match tok {
        Ok(tok) => (tok, Span::from(span)),
        Err(err) => panic!("lex error: {:?}", err),
    });
    let tok_stream = Stream::from_iter(tokens).spanned(Span::from(src.len()..src.len()));
    reader().parse(tok_stream).into_output_errors()
}

mod tests {
    // use super::Reader;

    // #[test]
    // fn read_int() {
    //     let src = "42";
    //     let mut reader = Reader::new(src);
    //     let (root, errs) = reader.read();
    //     if !errs.is_empty() {
    //         panic!("{:?}", errs);
    //     }
    //     insta::assert_debug_snapshot!(root);
    // }

    // #[test]
    // fn read_list() {
    //     let src = "(1 2 3)";
    //     let mut reader = Reader::new(src);
    //     let (root, errs) = reader.read();
    //     if !errs.is_empty() {
    //         panic!("{:?}", errs);
    //     }
    //     insta::assert_debug_snapshot!(root);
    // }
}
