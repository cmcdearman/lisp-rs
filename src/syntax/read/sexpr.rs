use num_rational::Rational64;

use crate::util::{intern::InternedString, list::List, node::SrcNode};
use std::fmt::{Debug, Display};

#[derive(Debug, Clone, PartialEq)]
pub struct Root {
    pub sexprs: Vec<SrcNode<Sexpr>>,
}

impl Display for Root {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        for s in self.clone().sexprs {
            writeln!(f, "{}", s.inner())?;
        }
        Ok(())
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Sexpr {
    Atom(SrcNode<Atom>),
    Cons(List<SrcNode<Sexpr>>),
}

impl Display for Sexpr {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self.clone() {
            Sexpr::Atom(a) => write!(f, "{}", a.inner()),
            Sexpr::Cons(l) => {
                write!(f, "(")?;
                for (i, s) in l.enumerate() {
                    if i > 0 {
                        write!(f, " ")?;
                    }
                    write!(f, "{}", s.inner())?;
                }
                write!(f, ")")
            }
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Atom {
    Symbol(InternedString),
    Number(Rational64),
    String(InternedString),
}

impl Display for Atom {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Atom::Symbol(s) => write!(f, "{}", s),
            Atom::Number(n) => write!(f, "{}", n),
            Atom::String(s) => write!(f, "{:?}", s),
        }
    }
}

// Sexpr format derive
// (+ 1 2)
// Root 0..7
//   Cons 0..7
//     Atom 0..1
//       Symbol 0..1
//         "+"
//     Atom 2..3
//       Lit 2..3
//         Int 2..3
//           1
//     Atom 4..5
//       Lit 4..5
//         Int 4..5
//           2
