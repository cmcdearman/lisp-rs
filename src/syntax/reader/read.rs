use super::{
    sexpr::{Atom, Pair, Root, Sexpr},
    token::Token,
};
use crate::util::{intern::InternedString, node::SrcNode, span::Span};
use chumsky::{
    extra,
    input::{Stream, ValueInput},
    prelude::{Input, Rich},
    primitive::just,
    recursive::recursive,
    select, IterParser, Parser,
};
use logos::Logos;

pub type ReadError<'a> = Rich<'a, Token, Span>;

fn sexpr_reader<'a, I: ValueInput<'a, Token = Token, Span = Span>>(
) -> impl Parser<'a, I, Sexpr, extra::Err<Rich<'a, Token, Span>>> {
    recursive(|sexpr| {
        let atom = select! {
            Token::Symbol(name) => Atom::Symbol(InternedString::from(name)),
            Token::Number(n) => Atom::Number(n),
            Token::String(s) => Atom::String(s),
        }
        .map_with_span(SrcNode::new)
        .map(Sexpr::Atom);

        let quote = just(Token::Quote)
            .map_with_span(SrcNode::new)
            .then(sexpr.clone())
            .map(|(q, sexpr)| {
                let quote = Sexpr::Atom(SrcNode::new(
                    Atom::Symbol(InternedString::from("quote")),
                    q.span(),
                ));
                Sexpr::from_iter(vec![quote, sexpr].into_iter().rev())
            });

        let quasiquote = just(Token::Backquote)
            .map_with_span(SrcNode::new)
            .then(sexpr.clone())
            .map(|(q, sexpr)| {
                let quote = Sexpr::Atom(SrcNode::new(
                    Atom::Symbol(InternedString::from("quasiquote")),
                    q.span(),
                ));
                Sexpr::from_iter(vec![quote, sexpr].into_iter().rev())
            });

        let unquote = just(Token::Comma)
            .map_with_span(SrcNode::new)
            .then(sexpr.clone())
            .map(|(q, sexpr)| {
                let quote = Sexpr::Atom(SrcNode::new(
                    Atom::Symbol(InternedString::from("unquote")),
                    q.span(),
                ));
                Sexpr::from_iter(vec![quote, sexpr].into_iter().rev())
            });

        let dot = sexpr
            .clone()
            .repeated()
            .collect::<Vec<_>>()
            .then_ignore(just(Token::Period))
            .then(sexpr.clone())
            .map(|(list, tail)| {
                list.into_iter().rev().fold(tail, |acc, next| {
                    Sexpr::Pair(SrcNode::new(
                        Pair::new(next.clone(), acc.clone()),
                        acc.span(),
                    ))
                })
            })
            .delimited_by(just(Token::LParen), just(Token::RParen));

        let list = sexpr
            .repeated()
            .collect::<Vec<_>>()
            .map(|sexprs| Sexpr::from_iter(sexprs.into_iter().rev()))
            .delimited_by(just(Token::LParen), just(Token::RParen));

        atom.or(list).or(quote).or(quasiquote).or(unquote).or(dot)
    })
}

fn reader<'a, I: ValueInput<'a, Token = Token, Span = Span>>(
) -> impl Parser<'a, I, SrcNode<Root>, extra::Err<Rich<'a, Token, Span>>> {
    sexpr_reader()
        .repeated()
        .collect()
        .map(Root)
        .map_with_span(SrcNode::new)
}

pub fn read<'src>(src: &'src str) -> (Option<SrcNode<Root>>, Vec<ReadError<'src>>) {
    let tokens = Token::lexer(&src).spanned().map(|(tok, span)| match tok {
        Ok(tok) => (tok, Span::from(span)),
        Err(err) => panic!("lex error: {:?}", err),
    });
    let tok_stream = Stream::from_iter(tokens).spanned(Span::from(src.len()..src.len()));
    reader().parse(tok_stream).into_output_errors()
}

mod tests {
    use crate::syntax::reader::read::read;

    #[test]
    fn read_int() {
        let src = "42";
        let (root, errs) = read(src);
        if !errs.is_empty() {
            panic!("{:?}", errs);
        }
        insta::assert_debug_snapshot!(root.unwrap());
    }

    #[test]
    fn read_list() {
        let src = "(1 2 3)";
        // (1 . (2 . (3 . ())))
        let (root, errs) = read(src);
        if !errs.is_empty() {
            panic!("{:?}", errs);
        }
        insta::assert_debug_snapshot!(root.unwrap());
    }

    #[test]
    fn read_quote() {
        let src = "'(1 2 3)";
        let (root, errs) = read(src);
        if !errs.is_empty() {
            panic!("{:?}", errs);
        }
        insta::assert_debug_snapshot!(root.unwrap());
    }

    #[test]
    fn read_quasiquote() {
        let src = "`(1 2 3)";
        let (root, errs) = read(src);
        if !errs.is_empty() {
            panic!("{:?}", errs);
        }
        insta::assert_debug_snapshot!(root.unwrap());
    }

    #[test]
    fn read_unquote() {
        let src = ",(1 2 3)";
        let (root, errs) = read(src);
        if !errs.is_empty() {
            panic!("{:?}", errs);
        }
        insta::assert_debug_snapshot!(root.unwrap());
    }

    #[test]
    fn read_quasi_unquote() {
        let src = "`(1 ,(+ 1 1) 3)";
        let (root, errs) = read(src);
        if !errs.is_empty() {
            panic!("{:?}", errs);
        }
        insta::assert_debug_snapshot!(root.unwrap());
    }

    #[test]
    fn read_dot() {
        let src = "(1 . 2)";
        let (root, errs) = read(src);
        if !errs.is_empty() {
            panic!("{:?}", errs);
        }
        insta::assert_debug_snapshot!(root.unwrap());
    }

    #[test]
    fn read_list_dot() {
        let src = "(1 2 . 3)";
        let (root, errs) = read(src);
        if !errs.is_empty() {
            panic!("{:?}", errs);
        }
        insta::assert_debug_snapshot!(root.unwrap());
    }
}
