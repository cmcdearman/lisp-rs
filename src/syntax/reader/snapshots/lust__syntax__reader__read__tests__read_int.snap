---
source: src/syntax/reader/read.rs
expression: root.unwrap()
---
Root @ 0..2
  Atom @ 0..2
    Number(42)
